File: apps/web/lib/types/linewise.ts
// Domain types — mirrors packages/contracts/module/schemas.py

export type Line = 14 | 17 | 19

export type SlotKind = 'production' | 'changeover' | 'cleaning' | 'maintenance'

/** Can formats: 50 cl (1/2), 33 cl (1/3), 44 cl (2/5). */
export type Format = '1/2' | '1/3' | '2/5'

/** Provenance of a changeover edge weight. */
export type EdgeSource = 'teorico' | 'empirico' | 'hibrido' | 'ml'

/** Decomposition of a total changeover into its driving segments. */
export type ChangeoverSegment =
  | 'brand'
  | 'container'
  | 'cap'
  | 'primary_pack'
  | 'secondary_pack'
  | 'pallet'
  | 'product'
  | 'volume'
  | 'startup'
  | 'shutdown'

export interface Slot {
  id: string
  line: Line
  /** ISO timestamp */
  start: string
  end: string
  kind: SlotKind
  sku?: string
  /** Display label: SKU code + family */
  label?: string
  units?: number
  /** Expected OEE for this slot [0–1] */
  oee_expected?: number
  /** Actual OEE from S_real, only on real sequence */
  oee_actual?: number
  /** Hours of changeover time (for changeover slots) */
  changeover_h?: number
  /** Whether changeover edge came from ML or theoretical matrix */
  changeover_source?: 'ml' | 'hibrido' | 'teorico'
  /** SHAP-style top features driving changeover time */
  changeover_drivers?: ChangeoverDriver[]
  /** True when this production slot was injected by an urgent-demand replan. */
  is_urgent?: boolean
}

export interface ChangeoverDriver {
  feature: string
  impact_h: number
}

export interface Sequence {
  id: string
  week_id: string
  week_start: string
  week_end: string
  source: 'opt' | 'real' | 'replan'
  slots: Slot[]
}

export interface LineMetrics {
  line: Line
  oee: number
  h_productive: number
  h_changeover: number
  h_cleaning: number
  h_maintenance: number
  h_idle: number
  coverage: number
}

export interface SimulationReport {
  sequence_id: string
  oee_global: number
  oee_per_line: LineMetrics[]
  h_changes: number
  h_productive: number
  coverage: number
  /** Total wall-clock hours from first to last slot */
  makespan_h: number
  dropped_skus: DroppedSku[]
}

export interface DroppedSku {
  sku: string
  units_demanded: number
  units_dropped: number
  margin_lost: number
  reason: string
}

export interface Inefficiency {
  id: string
  from_sku: string
  to_sku: string
  line: Line
  /** Average observed changeover time (hours) */
  avg_changeover_h: number
  /** Theoretical / expected changeover time (hours) */
  theoretical_h: number
  /** Extra hours lost beyond theoretical */
  avg_loss_h: number
  n_occurrences: number
  /** Human-readable explanation */
  reason: string
  /** Example week where this was worst */
  example_week?: string
}

export interface DeltaMetrics {
  oee_pp: number              // percentage-point delta (opt − real)
  h_changes_saved: number
  h_productive_gained: number
  coverage_delta: number
}

export type PerturbationKind = 'urgent_demand' | 'breakdown' | 'maintenance_shift'

export interface UrgentDemandPayload {
  sku: string
  units: number
  due_day: string
  priority: 'high' | 'medium'
}

export interface BreakdownPayload {
  line: Line
  start: string
  duration_h: number
}

export interface MaintenanceShiftPayload {
  line: Line
  original_day: string
  new_day: string
  duration_h: number
}

export interface Perturbation {
  kind: PerturbationKind
  payload: UrgentDemandPayload | BreakdownPayload | MaintenanceShiftPayload
}

export interface ReplanResult {
  sequence: Sequence
  simulation: SimulationReport
  dropped_skus: DroppedSku[]
}

export interface SkuMaster {
  sku: string
  family: string
  format: '1/2' | '1/3' | '2/5'
  brand: string
  margin_eur_per_unit: number
}

File: services/api/app/schemas/linewise.py
"""Pydantic schemas for the /linewise/* endpoints.

Field names follow the LINEWISE_API_CONTRACT.md spec exactly.
Percentages are fractions in [0, 1]. Line IDs are int literals 14/17/19.
"""

from __future__ import annotations

from typing import Literal, Optional

from pydantic import BaseModel


# ---------------------------------------------------------------------------
# Shared primitives
# ---------------------------------------------------------------------------

LineId = Literal[14, 17, 19]
SlotKind = Literal["production", "changeover", "cleaning", "maintenance"]
ChangeoverSourceApi = Literal["ml", "hibrido", "teorico"]


class ChangeoverDriver(BaseModel):
    feature: str
    impact_h: float


# ---------------------------------------------------------------------------
# Weeks
# ---------------------------------------------------------------------------

class WeekOption(BaseModel):
    id: str
    label: str
    range: str
    source: Literal["demo", "historical"]
    reason: str
    production_rows: Optional[int] = None
    sku_count: Optional[int] = None
    units: Optional[int] = None
    avg_oee: Optional[float] = None
    downtime_h: Optional[float] = None


# ---------------------------------------------------------------------------
# Sequence / Slots
# ---------------------------------------------------------------------------

class Slot(BaseModel):
    id: str
    line: LineId
    start: str
    end: str
    kind: SlotKind
    sku: Optional[str] = None
    label: Optional[str] = None
    units: Optional[int] = None
    oee_expected: Optional[float] = None
    oee_actual: Optional[float] = None
    changeover_h: Optional[float] = None
    changeover_source: Optional[ChangeoverSourceApi] = None
    changeover_drivers: Optional[list[ChangeoverDriver]] = None


class Sequence(BaseModel):
    id: str
    week_id: str
    week_start: str
    week_end: str
    source: Literal["opt", "real", "replan"]
    slots: list[Slot]


# ---------------------------------------------------------------------------
# Simulation reports
# ---------------------------------------------------------------------------

class LineMetrics(BaseModel):
    line: LineId
    oee: float
    h_productive: float
    h_changeover: float
    h_cleaning: float
    h_maintenance: float
    h_idle: float
    coverage: float


class DroppedSku(BaseModel):
    sku: str
    units_demanded: int
    units_dropped: int
    margin_lost: float
    reason: str


class SimulationReport(BaseModel):
    sequence_id: str
    oee_global: float
    oee_per_line: list[LineMetrics]
    h_changes: float
    h_productive: float
    coverage: float
    makespan_h: float
    dropped_skus: list[DroppedSku]


# ---------------------------------------------------------------------------
# Delta
# ---------------------------------------------------------------------------

class DeltaMetrics(BaseModel):
    oee_pp: float
    h_changes_saved: float
    h_productive_gained: float
    coverage_delta: float


# ---------------------------------------------------------------------------
# Compare bundle (endpoint 2)
# ---------------------------------------------------------------------------

class CompareBundle(BaseModel):
    week: WeekOption
    solution_id: str
    real_sequence: Sequence
    opt_sequence: Sequence
    real_simulation: SimulationReport
    opt_simulation: SimulationReport
    delta: DeltaMetrics


# ---------------------------------------------------------------------------
# Optimize plan (endpoint 3)
# ---------------------------------------------------------------------------

class ProductDemand(BaseModel):
    sku_id: str
    quantity_units: int


class PlanOptimizeRequest(BaseModel):
    products: list[ProductDemand]


class PlanGraphNode(BaseModel):
    id: str
    label: str
    line_id: LineId
    family: str
    volume_hl: float


class PlanGraphEdge(BaseModel):
    id: str
    source: str
    target: str
    hours: float
    path: Literal["opt", "baseline"]


class PlanOptimizeResponse(BaseModel):
    nodes: list[PlanGraphNode]
    edges: list[PlanGraphEdge]
    makespan_h: float
    h_saved: float
    coverage_pct: float
    dropped_skus: list[str]
    sequence: Sequence
